src/ripple/app/tx/impl/DepositPreauth.cpp

```diff
@@ -0,0 +1,195 @@
+            keylet::depositPreauth (ctx.tx[sfAccount], auth)))
+            return tecDUPLICATE;
+    }
+    else
+    {
+        // Verify that the Preauth entry they asked to remove is in the ledger.
+        AccountID const unauth {ctx.tx[sfUnauthorize]};
+        if (! ctx.view.exists (
+            keylet::depositPreauth (ctx.tx[sfAccount], unauth)))
+            return tecNO_ENTRY;
+    }
+    return tesSUCCESS;
+}
+
+TER
+DepositPreauth::doApply ()
+{
+    auto const sleOwner = view().peek (keylet::account (account_));
+
+    // Determine which operation we're performing: authorizing or unauthorizing.
+    if (ctx_.tx.isFieldPresent (sfAuthorize))
+    {
+        // A preauth counts against the reserve of the issuing account, but we
+        // check the starting balance because we want to allow dipping into the
+        // reserve to pay fees.
+        {
+            STAmount const reserve {view().fees().accountReserve (
+                sleOwner->getFieldU32 (sfOwnerCount) + 1)};
+
+            if (mPriorBalance < reserve)
+                return tecINSUFFICIENT_RESERVE;
+        }
+
+        // Preclaim already verified that the Preauth entry does not yet exist.
+        // Create and populate the Preauth entry.
+        AccountID const auth {ctx_.tx[sfAuthorize]};
+        uint256 const preauthIndex {getDepositPreauthIndex (account_, auth)};
+        auto slePreauth =
+            std::make_shared<SLE>(ltDEPOSIT_PREAUTH, preauthIndex);
+
+        slePreauth->setAccountID (sfAccount, account_);
+        slePreauth->setAccountID (sfAuthorize, auth);
+        view().insert (slePreauth);
+
+        auto viewJ = ctx_.app.journal ("View");
+        auto const page = dirAdd (view(), keylet::ownerDir (account_),
+            slePreauth->key(), false, describeOwnerDir (account_), viewJ);
+
+        JLOG(j_.trace())
+            << "Adding DepositPreauth to owner directory "
+            << to_string (slePreauth->key())
+            << ": " << (page ? "success" : "failure");
+
+        if (! page)
+            return tecDIR_FULL;
+
+        slePreauth->setFieldU64 (sfOwnerNode, *page);
+
+        // If we succeeded, the new entry counts against the creator's reserve.
+        adjustOwnerCount (view(), sleOwner, 1, viewJ);
+    }
+    else
+    {
+        // Verify that the Preauth entry they asked to remove is
+        // in the ledger.
+        AccountID const unauth {ctx_.tx[sfUnauthorize]};
+        uint256 const preauthIndex {getDepositPreauthIndex (account_, unauth)};
+        auto slePreauth = view().peek (keylet::depositPreauth (preauthIndex));
+
+        if (! slePreauth)
+        {
+            // Error should have been caught in preclaim.
+            JLOG(j_.warn()) << "Selected DepositPreauth does not exist.";
+            return tecNO_ENTRY;
+        }
+
+        auto viewJ = ctx_.app.journal ("View");
+        std::uint64_t const page {(*slePreauth)[sfOwnerNode]};
+        TER const ter {dirDelete (view(), true, page,
```

Comment:
For new code, prefer `view().dirRemove`

Reply:
Interesting.  `ApplyView::dirRemove()` returns a `bool`.  `dirDelete()` returns a `TER` which is either a `tefBAD_LEDGER` or a `tesSUCCESS`.  I'm in the transactor's `doApply()`, so I'd _prefer_ to return a `tec` code over a `tef`.  On the other hand, this should really only happen if the ledger is corrupt.  So returning a `tef` may be okay.  And it would keep me aligned with all the other places that call `dirDelete()`.

Would you prefer that I return a `tefBAD_LEDGER`, or a `tecINTERNAL` or a `tecFAILED_PROCESSING` if I'm unable to remove the directory entry?  Thanks.

Reply:
BTW, should I add a JIRA item to replace all occurrences of `dirAdd()` and `dirDelete()` with `ApplyView::dirInsert()` and `ApplyView::dirRemove()`?

Reply:
A failure in that code path would really be an exceptional circumstance. I'd be fine with returning `tefBAD_LEDGER` or a `tecFAILED_PROCESSING`.

Instead of a JIRA ticket, please create a GitHub Issue. Please be sure to note that `dirAdd` has two modes of operation, controlled by a `bool` argument: "insert" and "append" and they correspond to `ApplyView::dirInsert` and `ApplyView::dirAppend` respectively.